ecosystem/purescript-0.14-migration-guide.md

```diff
@@ -0,0 +1,622 @@
+
+The `Newtype` class exists to support convenient wrapping and unwrapping and to implement the combinators in the `Data.Newtype` module. The `Newtype` class now has `Coercible` as a superclass, which allows most combinators to now be implemented with `coerce` from the new `safe-coerce` library. This makes these combinators zero-cost for a performance gain.
+
+The `Newtype` class no longer has `wrap` and `unwrap` members. Instead, `wrap` and `unwrap` are synonyms for `coerce` and are exported as functions from the module. Most code does not need to change, but if you manually defined a `Newtype` instance in your code it will no longer compile.
+
+In addition, newtypes that do not have their constructors exported can no longer be unwrapped via the `Newtype` class. Most code will be unaffected, as giving a `Newtype` instance to a newtype with hidden constructors was already considered bad practice and usually accidental.
+
+**To fix:**
+
+Most `Newtype` instances are derived and will continue to work. If you manually defined a `Newtype` instance for a type, however, you can simply remove the `wrap` and `unwrap` definitions:
+
+```diff
++ instance newtypeAdditive :: Newtype (Additive a) a
+- instance newtypeAdditive :: Newtype (Additive a) a where
+-   wrap = Additive
+-   unwrap (Additive a) = a
+```
+
+Most newtypes either export their constructors and have a `Newtype` instance or hide their constructors and do not have the instance. However, if you have a newtype in your code which has hidden constructors and a `Newtype` instance then you will need to either:
+
+1. Remove the `Newtype` instance (a breaking change), or
+1. Export the newtype constructors so that `wrap` and `unwrap` can be used via the `Coercible` class.
+
+#### The `MonadZero` class has been deprecated.
+
+Implemented in [purescript-control#64](https://github.com/purescript/purescript-control/pull/64) with additional discussion in [purescript-control#62](https://github.com/purescript/purescript-control/issues/62) and [purescript-control#51](https://github.com/purescript/purescript-control/issues/51).
+
+The `MonadZero` class has been deprecated, as the additional law that it provides (`empty >>= f = empty`) is already covered by the classes it requires, namely `Monad` and `Alternative`.
+
+The `MonadZero` class was already little-used; in the core libraries it only shows up as a superclass of `MonadPlus` and as a constraint on the `guard` function. This release deprecates the class and it will be removed in the next release.
+
+Accordingly, the `guard` function has been moved from `Control.MonadZero` to `Control.Alternative` and the `MonadZero m` constraint has been replaced with `Monad m => Alternative m => ...`.
+
+**To fix:**
+
+If you are using the `guard` function, update your imports:
+
+```diff
+- import Control.MonadZero (guard)
++ import Control.Alternative (guard)
+```
+
+If you are using the `MonadZero m` constraint, replace it with `Monad m` and `Alternative m`:
```

Comment:
I think we should also say "if you're providing a MonadZero instance, consider removing it, as the class will be removed in a future release" here.